src/array.vala

```diff
@@ -92,28 +91,69 @@ public class Array<T>: Object
         return p;
     }
 
-    public unowned T
-    get_scalar(ssize_t [] index)
+    public void*
+    get_pointer ([CCode (array_length = false)] ssize_t[] index)
     {
-        return (T) ((uint8*) _data.get_data () + _offset_for_index (index));
+        return (uint8*) _data.get_data () + _offset_for_index (index);
     }
 
-    public ArrayIterator<T> iterator()
+    public Value
+    get_value ([CCode (array_length = false)] ssize_t[] index)
     {
-        return new ArrayIterator<T>(this);
+        var _value = Value (scalar_type);
+        if (_value.fits_pointer ()) {
+            Memory.copy (_value.peek_pointer (), get_pointer (index), scalar_size);
+        } else if (scalar_type == Type.BOXED) {
+            _value.set_boxed (get_pointer (index));
+        } else if (scalar_type == typeof (string)) {
+            _value.set_string ((string) get_pointer (index));
```

Comment:
The good thing if you count the `\0` is that you always have a null-terminator when you extract the memory (thus no copy with `(string) get_pointer ()`). 

If we pad with `\0`, we need to consider the case where the length of the string equals `instance_size` and make a copy.

I agree to pad with `\0` because it can avoid us systematic zeroing in `set_pointer`. For large strings, we could have a binary strategy where we pad only the end of the string until the first zero we encounter.